support-lib/cpp/Future.hpp

```diff
@@ -0,0 +1,317 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <optional>
+#include <condition_variable>
+#include <mutex>
+#include <cassert>
+
+#if defined(DJINNI_FUTURE_COROUTINE_SUPPORT)
+#if __has_include(<coroutine>)
+    #include <coroutine>
+    namespace snapchat::djinni::detail {
+        template <typename Promise = void> using CoroutineHandle = std::coroutine_handle<Promise>;
+    }
+#elif __has_include(<experimental/coroutine>)
+    #include <experimental/coroutine>
+    namespace snapchat::djinni::detail {
+        template <typename Promise = void> using CoroutineHandle = std::experimental::coroutine_handle<Promise>;
+    }
+#endif
+#endif
+
+namespace snapchat::djinni {
+namespace detail {
+
+// A wrapper object to support both void and non-void result types in
+// Promise/Future
+template <typename T>
+struct ValueHolder {
+    using type = T;
+    std::optional<T> value;
+    T getValueUnsafe() const {return *value;}
+};
+template <>
+struct ValueHolder<void> {
+    using type = bool;
+    std::optional<bool> value;
+    void getValueUnsafe() const {}
+};
+
+template<typename T>
+struct SharedState;
+
+// A simple type erased function container. It would be nice if std::function<>
+// supports move only lambdas.
+template <typename T>
+struct ValueHandlerBase {
+    virtual ~ValueHandlerBase() = default;
+    virtual void call(const std::shared_ptr<SharedState<T>>&) = 0;
+};
+template <typename T, typename F>
+class ValueHandler : public ValueHandlerBase<T> {
+public:
+    ValueHandler(F&& f) : _f(std::forward<F>(f)) {}
+    ValueHandler(ValueHandler&& other) : _f(std::move(other._f)) {}
+    void call (const std::shared_ptr<SharedState<T>>& s) override {
+        _f(s);
+    }
+private:
+    std::decay_t<F> _f;
+};
+template<typename T, typename FUNC>
+static auto createValueHandler(FUNC&& f) {
+    return std::make_unique<ValueHandler<T, FUNC>>(std::forward<FUNC>(f));
+}
+
+// The shared state object that links the promise and future objects
+template<typename T>
+struct SharedState: ValueHolder<T> {
+    std::condition_variable cv;
+    std::mutex mutex;
+    std::exception_ptr exception;
+    std::unique_ptr<ValueHandlerBase<T>> handler;
+
+    bool isReady() const {
+        return this->value.has_value() || exception != nullptr;
+    }
+};
+
+template<typename T>
+using SharedStatePtr = std::shared_ptr<SharedState<T>>;
+}
+
+template <typename T>
+class Future;
```

Comment:
are we ending the detail namespace on line 99 and starting again after this, just for the forward decl here?  can we just declare this above line 41, or should this be part of another header?